lib/log_download.py

```diff
@@ -0,0 +1,139 @@
+import datetime
+from dateutil.rrule import rrule, DAILY
+from datetime import timedelta
+from BeautifulSoup import BeautifulSoup
+from logging.config import fileConfig
+import urllib2 as ulib
+import logging
+import re
+import time
+import os
+
+# load the logging configuration
+fileConfig('logging_config.ini')
+
+# Global variables
+months = ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sept", "Oct", "Nov", "Dec"]
+UBUNTU_ENDPOINT = "https://irclogs.ubuntu.com/{}/{month:02d}/{day:02d}"
+UBUNTU_CHANNEL_ENDPOINT = "https://irclogs.ubuntu.com/{}/{month:02d}/{day:02d}/%23{channel}"
+SCUMMVM_ENDPOINT = "http://www.enderboi.com/log.php?log=scummvm.log.{day:02d}{month}{year}&format=text"        
+    
+def date_separate(dt):
+    """
+    Args:
+        dt (date object) : A valid date object
+    
+    Returns:
+        day  (int) : day corresponding to input date object
+        month(int) : month corresponding to input date object
+        year (int) : year corresponding to input date object
+    """
+    return dt.day, dt.month, dt.year
+
+def save_log(log, save_dir, dt, name):
+    """
+    Args:
+        log      (str) : chat log as text
+        save_dir (str) : location of directory to save the log
+        dt       (date): date corresponding to the log
+        name     (str) : name of the file to be saved
+    """
+    
+    if not (os.path.exists(save_dir) and os.path.isdir(save_dir)):
+        logging.info("Save Directory does exist. Creating a new directory {}".format(save_dir))
+        os.makedirs(save_dir)
+    
+    day,month,year = date_separate(dt)
+    
+    day_dir = os.path.join(save_dir, "{}/{month:02d}/{day:02d}".format(year, month=month, day=day))
+    
+    if not os.path.isdir(day_dir):
+        logging.info("Day log directory does not exist. Creating log directory {}".format(day_dir))
+        os.makedirs(day_dir)
+        
+    with open(os.path.join(day_dir, name), "w+") as f:
+        f.write(log)
+        
+def send_request(url):
+    """
+    Args:
+        url (str) - address to fetch log from
+    
+    Returns:
+        The response(log) received from the server.
+    """
+    
+    response = ulib.urlopen(url)
+    
+    status_code = response.getcode()
+    
+    if status_code != 200:
+        logging.warning("Error while sending download request to {} Returned code {}".format(url, status_code))
+    else:
+        logging.info("Log fetched successfully for {}".format(url))
+    
+    res = response.read()
+    time.sleep(1)
+    return res
+    
+
+def ubuntu_url(start_date, end_date):
+    """
+    Args:
+        start_date (date object): Starting date from which logs need to be fetched 
+        end_date (date object) : Last date for which logs need to be fetched
+    Returns:
+        Yields channel name, current_date, and url at which log for returned
+        channel and current_date is present.
+    """
+    
+    for current_date in list(rrule(freq=DAILY, dtstart=start_date, until=end_date)):
```

Comment:
This can be simply written as `for current_date in rrule(freq=DAILY, dtstart=start_date, until=end_date):`
rrule returns an iterator. so no need of converting it to a list object